features/topology-map/plugins/org.opennms.features.topology.plugins.topo.graphml/src/main/java/org/opennms/features/topology/plugins/topo/graphml/status/GraphMLPropagateVertexStatusProvider.java

```diff
@@ -0,0 +1,195 @@
+ * by the Free Software Foundation, either version 3 of the License,
+ * or (at your option) any later version.
+ *
+ * OpenNMS(R) is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with OpenNMS(R).  If not, see:
+ *      http://www.gnu.org/licenses/
+ *
+ * For more information contact:
+ *     OpenNMS(R) Licensing <license@opennms.org>
+ *     http://www.opennms.org/
+ *     http://www.opennms.com/
+ *******************************************************************************/
+
+package org.opennms.features.topology.plugins.topo.graphml.status;
+
+import com.google.common.collect.HashMultimap;
+import com.google.common.collect.Iterables;
+import com.google.common.collect.Maps;
+import com.google.common.collect.Multimap;
+import org.opennms.features.topology.api.topo.Criteria;
+import org.opennms.features.topology.api.topo.Status;
+import org.opennms.features.topology.api.topo.StatusProvider;
+import org.opennms.features.topology.api.topo.VertexProvider;
+import org.opennms.features.topology.api.topo.VertexRef;
+import org.opennms.features.topology.plugins.topo.graphml.GraphMLMetaTopologyProvider;
+import org.osgi.framework.BundleContext;
+import org.osgi.framework.InvalidSyntaxException;
+import org.osgi.framework.ServiceReference;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+
+public class GraphMLPropagateVertexStatusProvider implements StatusProvider {
+
+    private static final Logger LOG = LoggerFactory.getLogger(GraphMLPropagateVertexStatusProvider.class);
+
+    private static class LoopDetectionCriteria extends Criteria {
+        private final Set<VertexRef> seen;
+
+        private LoopDetectionCriteria(final LoopDetectionCriteria that,
+                                      final Collection<VertexRef> vertices) {
+            this();
+            this.seen.addAll(that.seen);
+            this.seen.addAll(vertices);
+        }
+
+        public LoopDetectionCriteria() {
+            this.seen = new HashSet<>();
+        }
+
+        public boolean contains(final VertexRef vertex) {
+            return this.seen.contains(vertex);
+        }
+
+
+        @Override
+        public ElementType getType() {
+            return ElementType.VERTEX;
+        }
+
+        @Override
+        public String getNamespace() {
+            return null;
+        }
+
+        @Override
+        public boolean equals(final Object o) {
+            if (this == o) {
+                return true;
+            }
+            if (o == null || getClass() != o.getClass()) {
```

Comment:
Is there a particular reason why `getClass()` is used instead of `instanceof`. 
`instanceof` is less restrictive and may be the better choice.